File: AlgoDat/Python/mult_russe.py
def mult_russe(a, b) -> int:
    return _mult_russe(a, b, 0)


def _mult_russe(a, b, sum) -> int:
    if a == 1:
        return sum + b
    elif a % 2 == 1:
        return _mult_russe(a//2, b*2, sum+b)
    else:
        return _mult_russe(a//2, b*2, sum)


print(mult_russe(45, 19))

File: AlgoDat/Go/mult_russe.go
package main

import "fmt"

func main() {
	fmt.Println(mult_russe(45, 19))
}

func mult_russe(a int, b int) (result int) {
	return mult_russe_helper(a, b, 0)
}

func mult_russe_helper(a int, b int, sum int) (result int) {
	if a == 1 {
		result = sum + b
		return result
	} else if a%2 == 1 {
		return mult_russe_helper(a/2, b*2, sum+b)
	} else {
		return mult_russe_helper(a/2, b*2, sum)
	}
}
